File: 394-decode-string/394-decode-string.java
class Solution {
    int[] pair;
    char[] arr;
    
    public String decodeString(String s) {
        arr = s.toCharArray();
        
        pair = new int[arr.length];
        
        Stack<Integer> stack = new Stack<>();
        for(int i = 0; i < arr.length; i++) {
            char c = arr[i];
            if (c == '[') { stack.push(i); }
            if (c == ']') { pair[stack.pop()] = i; }
        }
        
        return helper(0, arr.length - 1);
    }
    
    String helper(int s, int e) {
        StringBuilder sb = new StringBuilder();
        int n = 0;
        
        for(int i = s; i <= e; i++) {
            char c = arr[i];
            
            if (Character.isDigit(c)) {
                n = n * 10 + Character.getNumericValue(c);
                continue;
            }
            
            if (c == '[') {
                String sub = helper(i + 1, pair[i] - 1);
                sb.append(sub.repeat(n));
                n = 0;
                i = pair[i];
                continue;
            }
            
            sb.append(c);
        }
        
        return sb.toString();
    }
}

File: 394-decode-string/394-decode-string.scala
object Solution {
    def decodeString(s: String): String = {
        var mul = 0
        var idx = 0
        var res: String = ""
        
        while(idx < s.length) {
            val c = s.charAt(idx)

            if (c.isDigit)  mul = mul * 10 + c.getNumericValue
            else if (c == '[') {
                var stack = 1
                var eidx = idx + 1
                
                while(stack != 0) {
                    if (s.charAt(eidx) == '[') stack += 1
                    else if (s.charAt(eidx) == ']') stack -= 1
                    if (stack != 0) eidx += 1
                }
                
                res = res.concat(decodeString(s.subSequence(idx + 1, eidx).toString) * mul)
                idx = eidx
                mul = 0
            } else res += c

            idx += 1
        }

        res
    }
}